packages/transaction-controller/src/TransactionController.ts

```diff
@@ -556,14 +621,117 @@ export class TransactionController extends BaseControllerV1<
+      this.pendingTransactionTracker.onStateChange();
       this.onBootCleanup();
     });
 
     this.onBootCleanup();
+    this.messagingSystem.subscribe(
+      'NetworkController:stateChange',
+      (_, patches) => {
+        if (this.enableMultichain) {
+          const networkClients = this.getNetworkClientRegistry();
+          patches.forEach(({ op, path }) => {
+            if (op === 'remove' && path[0] === 'networkConfigurations') {
+              const networkClientId = path[1] as NetworkClientId;
+              delete networkClients[networkClientId];
+            }
+          });
+
+          this.#refreshTrackingMap(networkClients);
+          this.#refreshEtherscanRemoteTransactionSources(networkClients);
+        }
+      },
+    );
+    if (this.enableMultichain) {
+      this.#initTrackingMap();
+    }
   }
 
+  setEnableMultichain(enableMultichain: boolean) {
+    this.enableMultichain = enableMultichain;
+    if (enableMultichain) {
+      this.#initTrackingMap();
+    }
+  }
+
+  /**
+   * Stops polling and removes listeners to prepare the controller for garbage collection.
+   */
+  destroy() {
+    this.#stopAllTracking();
+  }
+
+  #refreshEtherscanRemoteTransactionSources = (
+    networkClients: ReturnType<NetworkController['getNetworkClientRegistry']>,
+  ) => {
+    // this will be prettier when we have consolidated network clients with a single chainId:
+    // check if there are still other network clients using the same chainId
+    // if not remove the etherscanRemoteTransaction source from the map
+    const chainIdsInRegistry = new Set();
+    Object.values(networkClients).forEach((networkClient) =>
+      chainIdsInRegistry.add(networkClient.configuration.chainId),
+    );
+    const existingChainIds = Array.from(
+      this.etherscanRemoteTransactionSourcesMap.keys(),
+    );
+    const chainIdsToRemove = existingChainIds.filter(
+      (chainId) => !chainIdsInRegistry.has(chainId),
+    );
+
+    chainIdsToRemove.forEach((chainId) => {
+      this.etherscanRemoteTransactionSourcesMap.delete(chainId);
+    });
+  };
+
+  #refreshTrackingMap = (
+    networkClients: ReturnType<NetworkController['getNetworkClientRegistry']>,
+  ) => {
+    const networkClientIds = Object.keys(networkClients);
+    const existingNetworkClientIds = Array.from(this.trackingMap.keys());
+
+    // Remove tracking for NetworkClientIds that no longer exist
+    const networkClientIdsToRemove = existingNetworkClientIds.filter(
+      (id) => !networkClientIds.includes(id),
+    );
+    networkClientIdsToRemove.forEach((id) => {
+      this.#stopTrackingByNetworkClientId(id);
+    });
+
+    // Start tracking new NetworkClientIds from the registry
+    const networkClientIdsToAdd = networkClientIds.filter(
+      (id) => !existingNetworkClientIds.includes(id),
+    );
+    networkClientIdsToAdd.forEach((id) => {
+      this.#startTrackingByNetworkClientId(id);
+    });
+  };
+
+  #initTrackingMap = () => {
```

Comment:
Given the tracking map consists of many properties and methods, is it worth encapsulating it inside its own helper in the interest of modularity?

It may also simplify unit tests?

Reply:
Maybe `MultichainHelper`?

Reply:
> Given the tracking map consists of many properties and methods, is it worth encapsulating it inside its own helper in the interest of modularity?

I agree that this could make managing this large file easier and it would feel cleaner. I think we'd prefer to pursue that cleanup in a subsequent refactor since a) we will be needing to come back shortly and remove the logic for globally selected network and b) perhaps even before that we will need to revisit some of this code once we consolidate the network configurations so it maybe a bit hard to anticipate the exact right abstractions for this new helper class

> It may also simplify unit tests?

The way we've written the new test suite aims to be maximally implementation agnostic so while it would help with unit tests, I don't think the additional compartmentalization would change the core testing we've added here.

Reply:
I feel really strongly about this and would much prefer that a helper be created prior to initial merge, since I don't consider it purely a refactor in the interest of tidiness, but a crucial design pattern we have to maintain in a large controller such as this. It should tangibly simplify the code review process also and could potentially identity otherwise missed issues.

The multi-chain architecture we're using is only possible because we've previously modularised and abstracted the distinct portions of logic into the helper classes, which the tracking map can then manage and initialise per network client. It seems counterintuitive to then have the logic built on top of that, not be a helper itself and not continue in the same spirit of modularity it relies on.

I agree it won't affect the integration tests, but it would be a huge simplifying factor and stability increase for the unit tests since they can be targeted specifically at the new multi-chain logic. Similarly, it should simplify `TransactionController.test.ts` as we wouldn't require as many unit tests for the additional embedded logic.

There is no specific API needed so I don't believe it should be an excessively complex task, but we'd just move the tracking map state and as many of the new methods as we can into the new helper. Then we'd simply require public methods for any interaction required from the `TransactionController` itself, such as the existing methods and any event listeners.

Reply:
@matthewwalsh0 Are there any examples you can give on how this helper should look? Would it essentially be a wrapper around the tracking map itself?

Reply:
The best examples are the existing helpers so `IncomingTransactionHelper` and `PendingTransactionTracker`.

The core of it would be the tracking map so it would aim to encapsulate any maintenance to that using constructor callbacks to register the listeners, plus have public methods to extract the necessary helper instances like `NonceTracker` for example.

Very rough but potential methods could be:

- constructor({onStateChange, onNetworkControllerStateChange, getNetworkClientRegistry, getNetworkClientById, addPendingTransactionTrackerListeners, addIncomingTransactionHelperListeners})
- startIncomingTransactionPolling(networkClientIds)
- stopIncomingTransactionPolling(networkClientIds)
- getNonceLock(networkClientId)
- destroy

This should mean we can encapsulate the vast majority of new logic such as:

- refreshTrackingMap
- checkForPendingTransactionAndStartPolling
- stopTrackingByNetworkClientId
- startTrackingByNetworkClientId
- removeIncomingTransactionHelperListeners
- stopAllTracking

Very rough and probably many mistakes, but hopefully you see the intent and encapsulation we could achieve.

Reply:
An attempt at the MultichainHelper

https://github.com/MetaMask/core/pull/3896